{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Program to facilitate manual editing of COVID-19 sequence.\n",
    "# The program uses MAFFT and read base coverage to provide additional\n",
    "# evidence to users to investigate evidence support the resolution of\n",
    "# ambigous bases, predicted indels, and variable sites.\n",
    "# Written in Python3 and require Biopython and pysam packages to be installed\n",
    "\n",
    "#Usage statement: python COVID_seq_edits.py\n",
    "\n",
    "#import in packages needed for script\n",
    "\n",
    "from Bio import AlignIO\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from datetime import datetime\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "from io import StringIO\n",
    "import pysam, os, subprocess\n",
    "from pysam import VariantFile\n",
    "import numpy as np\n",
    "\n",
    "#Set variables for program\n",
    "nucs=[\"A\",\"C\",\"G\",\"T\",\"N\",\"-\",\"R\",\"Y\",\"S\",\"W\",\"K\",\"M\",\"B\",\"D\",\"H\",\"B\"]\n",
    "variants={}\n",
    "Debug=False\n",
    "\n",
    "def summarize_alignment(align):\n",
    "    NumN = align[1,:].seq.count('N')\n",
    "    print(\"Number of ambiguous bases associated with each reference nucleotide type:\")\n",
    "    print(\"A->N:\", int(align.substitutions['N','A']*2))\n",
    "    print(\"C->N:\", int(align.substitutions['N','C']*2))\n",
    "    print(\"G->N:\", int(align.substitutions['N','G']*2))\n",
    "    print(\"T->N:\", int(align.substitutions['N','T']*2))    \n",
    "    print(\"Total Number of Ambiguous bases in genome assembly:\", NumN)   \n",
    "    print(\"Total Length of alignment:\", align.get_alignment_length())\n",
    "    print(\"Total Number of Variable Sites:\", len(variants))\n",
    "    print(\"Total Number of Non-ambiguous Variable Sites:\", len(variants)-NumN) \n",
    "\n",
    "# Code adapted from \"Base count extractor\" by david.eyre@ndm.ox.ac.uk \n",
    "# Method to count number of bases from reads mapped to a specific position\n",
    "def Bases_At_Pos(samfile, pos, chromname, minbasequal, minmapqual):\n",
    "    'Return a string of the bases at that position.'\n",
    "    position = 0\n",
    "    coverage = 0\n",
    "    bases = \"\"\n",
    "    for pileupcolumn in samfile.pileup(reference=chromname, start=pos-1, end=pos):\n",
    "        if ((pileupcolumn.pos+1) >= pos and (pileupcolumn.pos+1) <= pos):\n",
    "            position = int(pileupcolumn.pos+1)\n",
    "            coverage = int(pileupcolumn.n)\n",
    "            for pileupread in pileupcolumn.pileups:\n",
    "                if (pileupread.is_del == 0):\n",
    "                    #print(pileupread.query_position)\n",
    "                    bases += pileupread.alignment.seq[pileupread.query_position]\n",
    "    return position, coverage, bases\n",
    "\n",
    "# Method to summarize pileup base coverage and make base prediction\n",
    "# Key points: this methods will make a nucleotide prediction IF total\n",
    "# read coverage is >100 and the percentage of a base is >85%. This is\n",
    "# somewhat arbitrary, so feel free to adjust as you see fit. Also, the\n",
    "# method will suggest to manually investigate ANY indel in the query \n",
    "# sequence. Low coverage variable sites are identified with \"???\".\n",
    "def base_prediction(line, variants):\n",
    "    global totalBases\n",
    "    global predictedNuc\n",
    "    for i in range(len(line)):\n",
    "        if(i>0):\n",
    "            totalBases+=int(line[i])\n",
    "    for i in range(len(line)):\n",
    "        if(i>0):\n",
    "            if(totalBases>=100):\n",
    "                if((int(line[i])/float(totalBases))>=0.85 and variants[int(line[0])][1].upper()!='-'):\n",
    "                    if(i==1):\n",
    "                        predictedNuc = 'A'\n",
    "                    elif(i==2):\n",
    "                        predictedNuc = 'C'\n",
    "                    elif(i==3):\n",
    "                        predictedNuc = 'G'\n",
    "                    elif(i==4):\n",
    "                        predictedNuc = 'T'\n",
    "                elif((int(line[i])/float(totalBases))<0.85 and predictedNuc==\"\"):\n",
    "                    predictedNuc = 'Investigate'\n",
    "                \n",
    "            elif(totalBases < 25):\n",
    "                predictedNuc=\"???\"\n",
    "            else:\n",
    "                predictedNuc = 'Investigate'\n",
    "\n",
    "#Create file objects and import genome sequences\n",
    "tempFile = open(\"Temp.fa\", 'w')\n",
    "if Debug:\n",
    "    reference=SeqIO.read(\"MN908947.3.fasta\", 'fasta')\n",
    "else:\n",
    "    reference=SeqIO.read(input(\"Reference file?\"), 'fasta')\n",
    "if Debug:\n",
    "    query=SeqIO.read(\"MC01975.medaka.consensus.fasta\", 'fasta')\n",
    "else:\n",
    "    query=SeqIO.read(input(\"Query file?\"), 'fasta')\n",
    "seq_list= [reference.upper(), query.upper()]\n",
    "#Combine sequences and write to a temporary file\n",
    "SeqIO.write(seq_list,tempFile, 'fasta')\n",
    "tempFile.close()\n",
    "#run MAFFT to generate multiple sequence alignment\n",
    "mafft_exe = \"/Users/jue9332/opt/anaconda3/bin/mafft\"\n",
    "inFile=\"Temp.fa\"\n",
    "mafft_cline = MafftCommandline(mafft_exe, input=inFile)\n",
    "stdout, stderr = mafft_cline()\n",
    "align = AlignIO.read(StringIO(stdout.upper()), \"fasta\")\n",
    "query_seq=align[1,:]\n",
    "edit_seq=query.seq.tomutable()\n",
    "subprocess.call([\"rm\", \"Temp.fa\"])\n",
    "for i in range(align.get_alignment_length()):\n",
    "    if(\"n\" in align[:,i] or \"-\" in align[:,i] or align[:,i][0]!=align[:,i][1]):\n",
    "        variants[i+1]=align[:,i]\n",
    "# Summarize the variant/variable information\n",
    "summary = input(\"Would you like to summarize data? Y/N\")\n",
    "while(summary!='Y' and summary!='N'):\n",
    "    summary=input(\"Please type either a Y or an N\")    \n",
    "if(summary=='Y'):\n",
    "    summarize_alignment(align)\n",
    "\n",
    "align_array = np.array([list(rec) for rec in align], np.str_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Allow you to generate a predicted variant call based off of the multiple sequence alignment and bam/read coverage\n",
    "prediction = input(\"Would you like to generate predicted base calls for variable sites? Y/N\")\n",
    "\n",
    "while(prediction!='Y' and prediction!='N'):\n",
    "    prediction=input(\"Please type either a Y or an N\")\n",
    "if(prediction=='Y'):\n",
    "\n",
    "    #numeric list of positions of variable sites in reference genome, numbered from zero\n",
    "    POSLIST = list(variants.keys())\n",
    "    POSLIST.sort()\n",
    "    \n",
    "    #path to bam files for extraction of base counts\n",
    "    BAMPATH = input('Location of folder where bam files for base count extraction:)')\n",
    "\n",
    "    #file name for bam file, without .bam suffix, these will be recycled for output name\n",
    "    file = input(\"File name of bam file without suffix to be used for read counting:\")\n",
    "\n",
    "    #output path\n",
    "    OUTPATH = ''\n",
    "\n",
    "    #sets reference genome ID/name; can change if using different reference\n",
    "    ref = \"MN908947.3\"\n",
    "    #loop over each of the bam files\n",
    "    inbampath = BAMPATH+file+'.bam'\n",
    "    #will move to next file name if can't find bam file\n",
    "    if not os.path.exists(inbampath): \n",
    "        #skip this file\n",
    "        print(\"BAM file not found\")\n",
    "    else:\n",
    "        print(\"Generating pileup data for %s...\" % (inbampath))\n",
    "    #will create bam index file if it doesn't exist\n",
    "    if not os.path.exists(inbampath+\".bai\"):\n",
    "        pysam.index(inbampath)\n",
    "    #open bam file with pySAM library\n",
    "    samfile = pysam.Samfile(inbampath, 'rb')\n",
    "    #open output file and write headers\n",
    "    with open (OUTPATH+file+'_variant_sites.txt', 'w') as f:\n",
    "        f.write('site\\tA\\tC\\tG\\tT\\n')\n",
    "        #for each position of interest write base counts\n",
    "        for pos in POSLIST:\n",
    "            hq_position, hq_coverage, hq_bases = Bases_At_Pos(samfile, int(pos), ref, 30.0, 30.0)\n",
    "            hq_basecounts = [ len([x for x in hq_bases if x == 'A']), len([x for x in hq_bases if x == 'C']), \\\n",
    "                             len([x for x in hq_bases if x == 'G']), len([x for x in hq_bases if x == 'T']) ]\n",
    "            if(len([i for i in hq_basecounts if i>0])>=1):\n",
    "                f.write('%s\\t%s\\n'%(hq_position, '\\t'.join([str(i) for i in hq_basecounts])))\n",
    "        f.close()\n",
    "print(\"Done with pileup...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Option to allow for guided editing process\n",
    "edit = input(\"Would you like to edit as you go? Y/N\")\n",
    "while(edit!=\"Y\" and edit!='N'):\n",
    "    edit = input(\"Please type either a Y or an N\")\n",
    "# Option to output data on all variable sites or just the ones that are recommended for investigation\n",
    "output = input(\"Would you like to output all results to a file? \\nOtherwise, it will just write items that need to be investigated to a file. Y/N\")\n",
    "while(output!=\"Y\" and output!='N'):\n",
    "    output = input(\"Please type either a Y or an N\")\n",
    "    \n",
    "lineNumber=0\n",
    "tracker=0\n",
    "prevSite=0\n",
    "\n",
    "# examine results from pileup of reads, summarize data and make recommendations on variable sites\n",
    "pile_out = open(file+\"_variant_sites.txt\", \"r\")\n",
    "fout = open(file+\"_output.txt\", 'w')\n",
    "fout.write(\"#Ref\\tVariant\\tSite\\tNumReads\\tA\\tC\\tG\\tT\\tPredictedCall\\n\")\n",
    "    \n",
    "logFile = open(file+\".log\", 'w')\n",
    "now=datetime.now()\n",
    "dt_string = now.strftime(\"%d/%m/%Y %H:%M:%S\")\n",
    "logOut = \"\"\"#Log of changes made to %s sequence.\n",
    "#Date and time = %s\n",
    "\"\"\" % (query.id, dt_string)\n",
    "logFile.write(logOut)    \n",
    "    \n",
    "for line in pile_out:    \n",
    "    if(lineNumber==0):\n",
    "        if(edit==\"Y\"):\n",
    "            print(\"Ref\\tVariant\\tSite\\tNumReads\\tA\\tC\\tG\\tT\\tPredictedCall\")\n",
    "    else:\n",
    "        totalBases=0\n",
    "        line=line.split()\n",
    "        predictedNuc = \"\"\n",
    "        base_prediction(line, variants)\n",
    "        OutString=\"%s\\t%s\\t%s\\t%d\\t%.2f\\t%.2f\\t%.2f\\t%.2f\\t%s\" % (variants[int(line[0])][0].upper(),\\\n",
    "                                                                  variants[int(line[0])][1].upper(), \\\n",
    "                                                                  line[0], totalBases, int(line[1])/totalBases,\\\n",
    "                                                              int(line[2])/totalBases,\\\n",
    "                                                              int(line[3])/totalBases,\\\n",
    "                                                              int(line[4])/totalBases,\\\n",
    "                                                         predictedNuc)       \n",
    "        if((int(line[0])-prevSite)==1):\n",
    "            tracker+=1\n",
    "        else:\n",
    "            tracker=0\n",
    "        if(tracker>=1):\n",
    "            OutString=OutString + \"\\nWARNING: Possible Indel; May Need Investigation\"\n",
    "        prevSite=int(line[0])\n",
    "        start=int(line[0])\n",
    "        if(edit==\"Y\"):\n",
    "            print(OutString)\n",
    "        if(edit=='Y' and predictedNuc!=\"???\"):\n",
    "            confirm = input(\"Confirm that you would like to change the %s at position %s: Y/N\" % (variants[int(line[0])][1].upper(), line[0]))\n",
    "            while(confirm!=\"Y\" and confirm!='N'):\n",
    "                confirm = input(\"Please type either a Y or an N\")\n",
    "            if(confirm=='Y'):\n",
    "                change = input(\"Would you like to change this call to the predicted %s? Y/N\" % predictedNuc)\n",
    "                while(change!=\"Y\" and change!='N'):\n",
    "                    change = input(\"Please type either a Y or an N\")\n",
    "                if(change=='Y'):\n",
    "                    print(\"BERFORE:\", align_array[1,start-1])\n",
    "                    align_array[1,start-1]=predictedNuc\n",
    "                    print(\"AFTER:\", align_array[1,start-1])\n",
    "                    logOut=\"%s at position %s -> %s\" % (variants[int(line[0])][1].upper(),start,predictedNuc)\n",
    "                    logFile.write(logOut + \"\\n\")\n",
    "                else:\n",
    "                    alt = input(\"Is there a different call you would like to change this position to? Y/N\")\n",
    "                    while(alt!=\"Y\" and alt!='N'):\n",
    "                        alt = input(\"Please type either a Y or an N\")\n",
    "                    if(alt==\"Y\"):\n",
    "                        newNuc = input(\"What other bases call would you like to make at this position? A, C, G, T, or N?\")\n",
    "                        while(newNuc not in nucs):\n",
    "                            newNuc = input(\"Please choose a known base call code.\")\n",
    "                        print(\"BERFORE:\", align_array[1,start-1])\n",
    "                        align_array[1,start-1]=newNuc\n",
    "                        print(\"AFTER:\", align_array[1,start-1])\n",
    "                        logOut=\"%s at position %s -> %s\" % (variants[int(line[0])][1].upper(),start,newNuc)\n",
    "                        logFile.write(logOut + \"\\n\")\n",
    "            elif(output=='Y'):\n",
    "                fout.write(OutString + '\\n')\n",
    "                \n",
    "        elif(edit=='Y' and output=='N' and (predictedNuc==\"Investigate\" or predictedNuc==\"???\")):\n",
    "            fout.write(OutString + '\\n')\n",
    "        elif(output=='Y'):\n",
    "            fout.write(OutString + '\\n')\n",
    "    lineNumber+=1\n",
    "pile_out.close()\n",
    "fout.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edit sequence \"by hand\" - i.e. target specific sites manually\n",
    "looper = True\n",
    "while(looper==True):\n",
    "\n",
    "    byhand = input(\"Are there any edits you wish to make by hand? Y/N\")\n",
    "    while(byhand!=\"Y\" and byhand!='N'):\n",
    "        byhand = input(\"Please type either a Y or an N\")\n",
    "    if(byhand==\"Y\"):\n",
    "        print(\"YOU MUST EDIT YOUR SEQUENCE FROM THE RIGHT [i.e. highest numbered position] TO THE LEFT [i.e. lowest numbered position]\")\n",
    "        start=int(input(\"Start site on reference sequence?\"))\n",
    "        action=input(\"Action to take? C(hange),D(elete),I(nsert), or E(nd)\")\n",
    "        while(action not in ['C','D', 'I','E']):\n",
    "            action=input(\"Unrecognized option entered: Action to take? C,D, or I\")\n",
    "        print(\"The current base call at position %s is %s\" % (start, align_array[1,start-1]))\n",
    "        if(action==\"C\"):\n",
    "            new=str(input(\"What would you like to change this position to?\"))\n",
    "            align_array[1,start-1]=new\n",
    "            logOut=\"%s at position %s -> %s\" % (variants[int(line[0])][1].upper(),start,new)\n",
    "            logFile.write(logOut + \"\\n\")\n",
    "        elif(action==\"D\"):\n",
    "            indel=int(input(\"How many bases would you like to delete?\"))\n",
    "            align_array=np.concatenate([align_array[:,:(start-1)],align_array[:,(start+indel-1):]],axis=1)\n",
    "            logOut=\"%s nucleotides were deleted from the query sequence starting at position %s\" % (indel,start)\n",
    "            logFile.write(logOut + \"\\n\")\n",
    "        elif(action==\"I\"):\n",
    "            # It will insert after the position you provide\n",
    "            nucl=input(\"What bases would you like to insert?\")\n",
    "            for i in nucl:\n",
    "                align_array=np.insert(align_array,start,i,axis=1) \n",
    "            logOut=\"%s was inserted into the query sequence starting at position %s\" % (nucl,start)\n",
    "            logFile.write(logOut + \"\\n\")\n",
    "        else:\n",
    "            continue\n",
    "    if(byhand==\"N\"):\n",
    "        looper=False\n",
    "logFile.close()\n",
    "\n",
    "#write final edited sequence to new file\n",
    "edit_seq=\"\"\n",
    "for nuc in align_array[1,:]:\n",
    "    edit_seq+=nuc\n",
    "fasta=open(FILES[0]+\".fasta\",\"w\")\n",
    "fasta.write(\">\"+query.id+\"\\n\"+edit_seq)\n",
    "fasta.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
